### PR TITLE
address #8361 - introduce hook caller wrappers that enable backward compat

### LINKED ISSUE #8361: resolve startpath/fspath parameter regressions
the introductin of various `startpath`/`fspath` hooks in the hookspecs

 as a non-optional and non-inmplied parameter is practically a api breaking change

back in 2018 i decoded to close the upstream issue in https://github.com/pytest-dev/pluggy/issues/15 as its pretty much insanely complex to alter the hooks where callers and callees are not under good control

the pr adding them is #8144

CC @bluetech 

i believe/fear the correct solution would be to deprecate the old hook names, introduce a new hook version and have invoking/implementing the old one trigger a deprecation warning

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
- [x] unittests
- [ ] tests or receivers with legacy args + deprecation
- [x] remove legacy path parameters from internal calls
- [x] add warning for hook invocations on missing prams and suggested drops

fixes #8361

### CODE REVIEW

**Comment by asottile** [comment]
change makes sense, probably need a test or two to demonstrate the behaviour (do we want a warning and to eventually migrate?)

**Comment by nicoddemus** [comment]
I also think this is a good approach to #8361. 👍

**Review by bluetech — CHANGES_REQUESTED** [review]
Thanks for looking at this.

As I mentioned, I don't think we really *need* to do this, but if we can have a way to deprecate, I guess it's OK.

I *think* wrapping at the `config.hook` level is not sufficient -- some of the hooks use an `FSHookProxy`  (`gethookproxy`) which uses the plugin manager directly. I think if possible it would be better to add this directly on the `PytestPluginManager`.

I'm also think this might have some non-negligible effect on performance but I didn't measure. Can do once it's further along.

File: src/_pytest/main.py [hunk 1]
```diff
@@ -551,7 +551,9 @@ def gethookproxy(self, fspath: "os.PathLike[str]"):
         remove_mods = pm._conftest_plugins.difference(my_conftestmodules)
         if remove_mods:
             # One or more conftests are not in use at this fspath.
-            proxy = FSHookProxy(pm, remove_mods)
+            from .config.compat import PathAwareHookProxy
+
+            proxy = PathAwareHookProxy(FSHookProxy(pm, remove_mods))
```
Comment (RonnyPfannschmidt) [inline src/_pytest/main.py:556]: @bluetech here i wrap that one

**Comment by RonnyPfannschmidt** [comment]
@bluetech i believe, in a followup i can even add some bits to ensure the cost can be removed if people opt out of legacy paths

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Overall great work, please see my comments!

File: doc/en/deprecations.rst [hunk 2]
```diff
@@ -19,6 +19,20 @@ Below is a complete list of all pytest features which are considered deprecated.
 :class:`PytestWarning` or subclasses, which can be filtered using :ref:`standard warning filters <warnings>`.
 
 
+``py.path.local`` arguments for hooks replaced with ``pathlib.Path``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In order to support the transition oto pathlib, the following hooks added additional arugments:
```
Comment (nicoddemus) [inline doc/en/deprecations.rst:25]:
```suggestion
In order to support the transition to :mod:`pathlib`, the following hooks now receive additional arguments:
```

File: doc/en/deprecations.rst [hunk 3]
```diff
@@ -19,6 +19,20 @@ Below is a complete list of all pytest features which are considered deprecated.
...
+
+*  ``pytest_ignore_collect(fspath: pathlib.Path)``
```
Comment (nicoddemus) [inline doc/en/deprecations.rst:27]:
```suggestion
*  :func:`pytest_ignore_collect(fspath: pathlib.Path) <_pytest.hookspec.pytest_ignore_collect>`
```

File: doc/en/deprecations.rst [hunk 4]
```diff
@@ -19,6 +19,20 @@ Below is a complete list of all pytest features which are considered deprecated.
...
+*  ``pytest_collect_file(fspath: pathlib.Path)``
```
Comment (nicoddemus) [inline doc/en/deprecations.rst:28]: *  :func:`pytest_collect_file(fspath: pathlib.Path) <_pytest.hookspec.pytest_collect_file>`

File: doc/en/deprecations.rst [hunk 5]
```diff
@@ -19,6 +19,20 @@ Below is a complete list of all pytest features which are considered deprecated.
...
+*  ``pytest_pycollect_makemodule(fspath: pathlib.Path)``
```
Comment (nicoddemus) [inline doc/en/deprecations.rst:29]:
```suggestion
*  :func:`pytest_collect_file(fspath: pathlib.Path) <_pytest.hookspec.pytest_collect_file>`
```

File: doc/en/deprecations.rst [hunk 6]
```diff
@@ -19,6 +19,20 @@ Below is a complete list of all pytest features which are considered deprecated.
...
+*  ``pytest_report_header(startpath: pathlib.Path)``
```
Comment (nicoddemus) [inline doc/en/deprecations.rst:30]:
```suggestion
*  :func:`pytest_collect_file(fspath: pathlib.Path) <_pytest.hookspec.pytest_collect_file>`
```

File: doc/en/deprecations.rst [hunk 7]
```diff
@@ -19,6 +19,20 @@ Below is a complete list of all pytest features which are considered deprecated.
...
+*  ``pytest_report_collectionfinish(startpath: pathlib.Path)``
```
Comment (nicoddemus) [inline doc/en/deprecations.rst:31]:
```suggestion
*  :func:`pytest_collect_file(fspath: pathlib.Path) <_pytest.hookspec.pytest_collect_file>`
```

File: doc/en/deprecations.rst [hunk 8]
```diff
@@ -19,6 +19,20 @@ Below is a complete list of all pytest features which are considered deprecated.
...
+
+The accompanying ``py.path.local`` based paths have been deprecated.
```
Comment (nicoddemus) [inline doc/en/deprecations.rst:33]:
```suggestion
The accompanying ``py.path.local`` based paths have been deprecated: plugins which manually invoke those hooks should only pass the new ``pathlib.Path`` arguments, and users should change their hook implementations to use the new ``pathlib.Path`` arguments.
```

File: src/_pytest/config/compat.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+            hook = getattr(self.__hook_caller, key)
+            path_var, fspath_var = imply_paths_hooks[key]
+
+            def fixed_hook(**kw):
+                path_value: Optional[Path] = kw.pop(path_var, None)
+                fspath_value: Optional[LEGACY_PATH] = kw.pop(fspath_var, None)
+                if fspath_value is not None:
+                    warnings.warn(
+                        HOOK_LEGACY_PATH_ARG.format(
+                            pylib_path_arg=fspath_var, pathlib_path_arg=path_var
```
[the remark this answers is not in the document]
Comment (RonnyPfannschmidt) [reply src/_pytest/config/compat.py:51]: fixed

**Review by asottile — APPROVED** [review]
looks good!  thanks for the good tests

**Review by nicoddemus — APPROVED** [review]

**Review by bluetech — APPROVED** [review]
LGTM!

**Pull request merged**